<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Ted Laderas" />

<meta name="date" content="2017-11-28" />

<title>Bayesian Network for Synthetic Data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Bayesian Network for Synthetic Data</h1>
<h4 class="author"><em>Ted Laderas</em></h4>
<h4 class="date"><em>2017-11-28</em></h4>



<p>In this file, I start generating the synthetic data using the <code>gRain</code> package.</p>
<div id="basic-synthetic-dataset" class="section level2">
<h2>Basic Synthetic Dataset</h2>
<p>The first variable we add is <code>age</code>. Then I add <code>smoking</code>, which is associated with hypertension.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">yn &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;N&quot;</span>,<span class="st">&quot;Y&quot;</span>)

<span class="co">#this is the true conditional prob table for HTNtreat given HTN status</span>
<span class="co">#condHTNtreat &lt;- t(HTNtreat/rowSums(HTNtreat))</span>

<span class="co">#I adjusted according to David's suggestions</span>
condHTNtreat &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">matrix</span>(<span class="dt">nrow=</span><span class="dv">2</span>,<span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">0</span>,<span class="dv">40</span>,<span class="dv">60</span>), <span class="dt">byrow =</span> <span class="ot">TRUE</span>, <span class="dt">dimnames=</span><span class="kw">list</span>(yn,yn)))

## Age probabilities are calculated from our cohort
ageLevels &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;0-20&quot;</span>,<span class="st">&quot;20-40&quot;</span>, <span class="st">&quot;40-55&quot;</span>, <span class="st">&quot;55-70&quot;</span>,<span class="st">&quot;70-90&quot;</span>)

ageProbMat &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="kw">system.file</span>(<span class="st">&quot;cpt&quot;</span>, <span class="st">&quot;age_probabilities.csv&quot;</span>, <span class="dt">package=</span><span class="st">&quot;cvdRiskData&quot;</span>))
ageProb &lt;-<span class="st"> </span>ageProbMat<span class="op">$</span>Freq
<span class="kw">names</span>(ageProb) &lt;-<span class="st"> </span>ageProbMat<span class="op">$</span>out

ageProb</code></pre></div>
<pre><code>##     (0,20]    (20,40]    (40,55]    (55,70]    (70,90] 
## 0.08744536 0.29710261 0.25755875 0.23272094 0.12517234</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>age, <span class="dt">values=</span>ageProb<span class="op">*</span><span class="dv">100</span>, <span class="dt">levels=</span>ageLevels)

## These conditional probabilities are derived from our cohort
condHTNAgeprob &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="kw">system.file</span>(<span class="st">&quot;cpt&quot;</span>, <span class="st">&quot;age_htn_cpt.csv&quot;</span>, <span class="dt">package=</span><span class="st">&quot;cvdRiskData&quot;</span>), <span class="dt">check.names =</span> <span class="ot">FALSE</span>)

## Here are the conditional age|htn probabilities
condHTNAgeprob</code></pre></div>
<pre><code>##       (0,20]   (20,40]   (40,55]  (55,70]   (70,90]
## 1 0.97665607 0.8516853 0.7206523 0.523628 0.3901981
## 2 0.02334393 0.1483147 0.2793477 0.476372 0.6098019</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rownames</span>(condHTNAgeprob) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)
<span class="kw">colnames</span>(condHTNAgeprob) &lt;-<span class="st"> </span>ageLevels

## Compile our CPTs so far
a.htn &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>htn<span class="op">+</span>age, <span class="dt">values=</span><span class="kw">as.matrix</span>(condHTNAgeprob), <span class="dt">levels=</span>yn)
htn.treat &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>treat<span class="op">+</span>htn, <span class="dt">values=</span>condHTNtreat, <span class="dt">levels=</span>yn)

##Smoking|age CPT - peaks at 25% smoking for 55-70 category
a.s &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>smoking<span class="op">+</span>age, <span class="dt">values=</span><span class="kw">c</span>(<span class="dv">95</span>,<span class="dv">5</span>,<span class="dv">90</span>,<span class="dv">10</span>,<span class="dv">85</span>,<span class="dv">15</span>,<span class="dv">75</span>,<span class="dv">25</span>,<span class="dv">90</span>,<span class="dv">10</span>),<span class="dt">levels=</span>yn)
##Smoking htn, if you smoke, you have 67% prob of having hypertension
s.h &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>htn<span class="op">+</span>smoking, <span class="dt">values=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">50</span>,<span class="dv">33</span>,<span class="dv">67</span>))</code></pre></div>
<p>Now we load in probabilities for race, which was calculated from our patient cohort.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pRaceMat &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="kw">system.file</span>(<span class="st">&quot;cpt&quot;</span>, <span class="st">&quot;race_probabilities.csv&quot;</span>, <span class="dt">package=</span><span class="st">&quot;cvdRiskData&quot;</span>))

pRace &lt;-<span class="st"> </span>pRaceMat<span class="op">$</span>Freq
<span class="kw">names</span>(pRace) &lt;-<span class="st"> </span>pRaceMat<span class="op">$</span>Var1

pRace</code></pre></div>
<pre><code>##       AmInd    Asian/PI  Black/AfAm       White 
## 0.005527018 0.191968947 0.053266109 0.749237925</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">race &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>race, <span class="dt">values=</span>pRace, <span class="dt">levels=</span><span class="kw">names</span>(pRace))

<span class="co">#These probabilities are estimated for US patients overall (from CDC report)</span>
bmiLevels &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;15-18&quot;</span>,<span class="st">&quot;18-25&quot;</span>, <span class="st">&quot;25-31&quot;</span>,<span class="st">&quot;31+&quot;</span>)
bmiProb &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="fl">0.16</span>, <span class="fl">0.68</span>, <span class="fl">0.10</span>, <span class="fl">0.06</span>)

bmi &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>bmi, <span class="dt">values=</span>bmiProb, <span class="dt">levels=</span>bmiLevels)

<span class="co">#conditional CPT for BMI given race</span>
<span class="co">#these values were estimated from CDC report</span>
<span class="co">#If you are Black/AfAm or White, have a higher </span>
<span class="co">#probability of being in a higher BMI category</span>
valuesBMI &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">97</span>, <span class="dv">3</span>, <span class="dv">98</span>, <span class="dv">1</span>, <span class="dv">98</span>, <span class="dv">2</span>, <span class="dv">99</span>, <span class="dv">1</span>,
            <span class="dv">88</span>, <span class="dv">12</span>, <span class="dv">92</span>,<span class="dv">8</span>, <span class="dv">92</span>, <span class="dv">8</span>, <span class="dv">96</span>, <span class="dv">4</span>,
            <span class="dv">55</span>, <span class="dv">45</span>, <span class="dv">70</span>, <span class="dv">30</span>, <span class="dv">70</span>, <span class="dv">30</span>, <span class="dv">85</span>, <span class="dv">15</span>,
            <span class="dv">20</span>, <span class="dv">80</span>, <span class="dv">33</span>, <span class="dv">66</span>, <span class="dv">33</span>, <span class="dv">66</span>, <span class="dv">65</span>, <span class="dv">35</span>)

##t2d.race.bmi is a joint cpt: p(t2d|race,bmi)
t2d.race.bmi &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>t2d<span class="op">|</span>race<span class="op">:</span>bmi, <span class="dt">values=</span><span class="kw">c</span>(valuesBMI), <span class="dt">levels=</span>yn)
<span class="co">#t2d.cvd &lt;- cptable(~cvd|t2d, values=c(95, 5, 5, 95), levels= yn)</span>

##incidence of smoking is 15%
smoking &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>smoking, <span class="dt">values=</span><span class="kw">c</span>(<span class="dv">85</span>, <span class="dv">15</span>), <span class="dt">levels=</span>yn)</code></pre></div>
</div>
<div id="selected-snps" class="section level2">
<h2>Selected SNPs</h2>
<p>These first 2 snps are on the same chromosome and so I will model them as having identical risk, and having both does not increase risk. I am limiting myself to the homozygous variants (excluding the heterozygous variants) in order to simplify things. I also only consider 6 genotypes overall.</p>
<p><a href="http://snpedia.com/index.php/Rs10757278">Rs10757278</a> (Chr 9, 22124478). (A;A) 0.78 x risk of heart disease, (A;G) 1.3 x risk, (G;G) 1.6x risk for heart disease. Co-morbidity: diabetes</p>
<p><a href="http://snpedia.com/index.php/Rs1333049">Rs1333049</a> (Chr 9, 22125504). (G;G) Normal (WT); (C;G) 1.5 increased risk; (C;C) 1.9x increased risk.</p>
<p>These last 2 snps are on different chromosomes for the first. Again, limiting to the homozygous cases to simplify things.</p>
<p><a href="http://snpedia.com/index.php/Rs4665058">Rs4665058</a>, 4x risk if have two copies of the SNP (A;A), 2 X (A;C) and (C;C) - wild type. Chr 2, 159333698</p>
<p><a href="https://www.snpedia.com/index.php/Rs8055236">rs8055236</a> (Chr 16, 83178793) (T;T) Normal (WT); (G;T) 1.9x risk, (G;G) 2.2x increased risk</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">raceNames &lt;-<span class="st"> </span><span class="kw">names</span>(pRace)

<span class="co">#Conversion table between aggregate genotype and SNP values</span>
snpLookup &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;0000&quot;</span> =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;AA&quot;</span>, <span class="st">&quot;CC&quot;</span>, <span class="st">&quot;CC&quot;</span>, <span class="st">&quot;TT&quot;</span>),
     <span class="st">&quot;0001&quot;</span> =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;AA&quot;</span>, <span class="st">&quot;CC&quot;</span>, <span class="st">&quot;CC&quot;</span>, <span class="st">&quot;GG&quot;</span>),
     <span class="st">&quot;0010&quot;</span> =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;AA&quot;</span>, <span class="st">&quot;CC&quot;</span>, <span class="st">&quot;AA&quot;</span>, <span class="st">&quot;TT&quot;</span>),
     <span class="st">&quot;1100&quot;</span> =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GG&quot;</span>, <span class="st">&quot;GG&quot;</span>, <span class="st">&quot;CC&quot;</span>, <span class="st">&quot;TT&quot;</span>),
     <span class="st">&quot;1110&quot;</span> =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GG&quot;</span>, <span class="st">&quot;GG&quot;</span>, <span class="st">&quot;AA&quot;</span>, <span class="st">&quot;GG&quot;</span>),
     <span class="st">&quot;1111&quot;</span> =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GG&quot;</span>, <span class="st">&quot;GG&quot;</span>, <span class="st">&quot;AA&quot;</span>, <span class="st">&quot;GG&quot;</span>))

snpNames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;rs10757278&quot;</span>, <span class="st">&quot;rs1333049&quot;</span>, <span class="st">&quot;rs4665058&quot;</span>, <span class="st">&quot;rs8055236&quot;</span>)

snpLookup &lt;-<span class="st"> </span><span class="kw">lapply</span>(snpLookup, <span class="cf">function</span>(x) {<span class="kw">names</span>(x) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;rs10757278&quot;</span>, <span class="st">&quot;rs1333049&quot;</span>, <span class="st">&quot;rs4665058&quot;</span>, <span class="st">&quot;rs8055236&quot;</span>); <span class="kw">return</span>(x)})

rs10757278race =<span class="st"> </span><span class="kw">matrix</span>(<span class="dt">nrow=</span><span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>, <span class="dt">data=</span><span class="kw">c</span>(<span class="dv">21</span>, <span class="dv">24</span>, <span class="dv">80</span>, <span class="dv">30</span>, 
                                                     <span class="dv">79</span>, <span class="dv">76</span>, <span class="dv">20</span>, <span class="dv">70</span>), 
                        <span class="dt">dimnames =</span> <span class="kw">list</span>( <span class="kw">c</span>(<span class="st">&quot;AA&quot;</span>,<span class="st">&quot;GG&quot;</span>),raceNames))<span class="op">/</span><span class="dv">100</span>

r1 &lt;-<span class="st"> </span>rs10757278race

<span class="co">#p(&quot;AA&quot;|race) = c(10, 10, 10, 10) </span>
<span class="co">#p(&quot;CC&quot;|race) = c(80,80,80,80)</span>

rs4665058race =<span class="st"> </span>
<span class="st">  </span><span class="kw">matrix</span>(<span class="dt">nrow=</span><span class="dv">2</span>, <span class="dt">data=</span> <span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">90</span>,<span class="dv">90</span>,<span class="dv">90</span>,<span class="dv">90</span>), <span class="dt">byrow=</span><span class="ot">TRUE</span>, 
         <span class="dt">dimnames=</span> <span class="kw">list</span>(<span class="kw">c</span>(<span class="st">&quot;AA&quot;</span>,<span class="st">&quot;CC&quot;</span>), raceNames ))<span class="op">/</span><span class="dv">100</span>

r2 &lt;-<span class="st"> </span>rs4665058race

<span class="co">#p(&quot;GG&quot;|race) = c(62, 65, 20, 62)</span>
<span class="co">#P(&quot;TT&quot;|race) = c(38, 35, 80, 38)</span>

rs8055236race =<span class="st"> </span><span class="kw">matrix</span>(<span class="dt">nrow=</span><span class="dv">2</span>, <span class="dt">byrow=</span><span class="ot">TRUE</span>, <span class="dt">data=</span><span class="kw">c</span>(<span class="dv">62</span>, <span class="dv">65</span>, <span class="dv">20</span>, <span class="dv">40</span>, 
                                                  <span class="dv">38</span>, <span class="dv">35</span>, <span class="dv">80</span>, <span class="dv">60</span>),
                       <span class="dt">dimnames =</span> <span class="kw">list</span>(<span class="kw">c</span>(<span class="st">&quot;GG&quot;</span>,<span class="st">&quot;TT&quot;</span>), raceNames))<span class="op">/</span><span class="dv">100</span>
r3 &lt;-<span class="st"> </span>rs8055236race

genotypes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;1111&quot;</span>, <span class="st">&quot;1110&quot;</span>, <span class="st">&quot;1100&quot;</span>, <span class="st">&quot;0010&quot;</span>, <span class="st">&quot;0001&quot;</span>, <span class="st">&quot;0000&quot;</span>)

probList &lt;-<span class="st"> </span><span class="kw">list</span>()
probList[[<span class="st">&quot;0000&quot;</span>]] &lt;-<span class="st"> </span>r1[<span class="dv">2</span>,] <span class="op">*</span><span class="st"> </span>r2[<span class="dv">2</span>,] <span class="op">*</span><span class="st"> </span>r3[<span class="dv">2</span>,]
probList[[<span class="st">&quot;1100&quot;</span>]] &lt;-<span class="st"> </span>r1[<span class="dv">1</span>,] <span class="op">*</span><span class="st"> </span>r2[<span class="dv">2</span>,] <span class="op">*</span><span class="st"> </span>r3[<span class="dv">2</span>,]
probList[[<span class="st">&quot;1110&quot;</span>]] &lt;-<span class="st"> </span>r1[<span class="dv">1</span>,] <span class="op">*</span><span class="st"> </span>r2[<span class="dv">1</span>,] <span class="op">*</span><span class="st"> </span>r3[<span class="dv">2</span>,]
probList[[<span class="st">&quot;0010&quot;</span>]] &lt;-<span class="st"> </span>r1[<span class="dv">2</span>,] <span class="op">*</span><span class="st"> </span>r2[<span class="dv">1</span>,] <span class="op">*</span><span class="st"> </span>r3[<span class="dv">2</span>,]
probList[[<span class="st">&quot;0001&quot;</span>]] &lt;-<span class="st"> </span>r1[<span class="dv">2</span>,] <span class="op">*</span><span class="st"> </span>r2[<span class="dv">2</span>,] <span class="op">*</span><span class="st"> </span>r3[<span class="dv">1</span>,]
probList[[<span class="st">&quot;1111&quot;</span>]] &lt;-<span class="st"> </span>r1[<span class="dv">1</span>,] <span class="op">*</span><span class="st"> </span>r2[<span class="dv">1</span>,] <span class="op">*</span><span class="st"> </span>r3[<span class="dv">1</span>,]

cptgenoRace &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind,probList)
cptgenoRace &lt;-<span class="st"> </span>cptgenoRace[genotypes,]

race.genotype &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>genotype<span class="op">|</span>race, <span class="dt">values=</span>cptgenoRace, <span class="dt">levels=</span>genotypes, <span class="dt">normalize=</span><span class="ot">TRUE</span>)

##Gender is unbalanced
gender &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>gender, <span class="kw">c</span>(<span class="dv">45</span>,<span class="dv">55</span>), <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;M&quot;</span>,<span class="st">&quot;F&quot;</span>))

##Two SNPs are associated with higher cholesterol
tcholStates &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;&lt;160&quot;</span>, <span class="st">&quot;160-199&quot;</span>, <span class="st">&quot;200-239&quot;</span>,<span class="st">&quot;240+&quot;</span>)
genotype.tchol &lt;-<span class="st"> </span><span class="kw">cptable</span>(<span class="op">~</span>tchol<span class="op">|</span>genotype, 
                          <span class="dt">values =</span> <span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>,<span class="dv">30</span>,<span class="dv">55</span>,
                                     <span class="dv">10</span>,<span class="dv">20</span>,<span class="dv">30</span>,<span class="dv">40</span>,
                                     <span class="dv">15</span>,<span class="dv">20</span>,<span class="dv">45</span>,<span class="dv">30</span>,
                                     <span class="dv">15</span>,<span class="dv">20</span>,<span class="dv">45</span>,<span class="dv">30</span>,
                                     <span class="dv">30</span>,<span class="dv">60</span>,<span class="dv">5</span>,<span class="dv">5</span>,
                                     <span class="dv">30</span>,<span class="dv">60</span>,<span class="dv">5</span>,<span class="dv">5</span>), <span class="dt">levels=</span>tcholStates)</code></pre></div>
<p>Finally, we compile our Bayesian Network and save it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plist &lt;-<span class="st"> </span><span class="kw">compileCPT</span>(<span class="kw">list</span>(a, a.htn, htn.treat, smoking, s.h, race,
                         bmi, t2d.race.bmi, race.genotype, 
                         genotype.tchol,
                         gender))

cvd_bayes_net &lt;-<span class="st"> </span><span class="kw">compile</span>(<span class="kw">grain</span>(plist))
devtools<span class="op">::</span><span class="kw">use_data</span>(cvd_bayes_net, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## Saving cvd_bayes_net as cvd_bayes_net.rda to /Users/roshi/Code/cvdRiskData/data</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
